heron/common/src/cpp/config/topology-config-helper.cpp

```diff
@@ -399,17 +399,105 @@ bool TopologyConfigHelper::DropTuplesUponBackpressure(const proto::api::Topology
+}
+
+// Return component level config
+void TopologyConfigHelper::GetComponentConfig(const proto::api::Topology& _topology,
+                                              const sp_string& _component_name,
+                                              std::map<sp_string, sp_string>& retval) {
+  // We are assuming component names are unique and returning the config
+  // of the first spout or bolt found with the name.
+  for (sp_int32 i = 0; i < _topology.spouts_size(); ++i) {
+    if (_topology.spouts(i).comp().name() == _component_name) {
+      if (_topology.spouts(i).comp().has_config()) {
+        const proto::api::Config& config = _topology.spouts(i).comp().config();
+        ConvertConfigToKVMap(config, retval);
+      }
+      return;
+    }
+  }
+
+  for (sp_int32 i = 0; i < _topology.bolts_size(); ++i) {
+    if (_topology.bolts(i).comp().name() == _component_name) {
+      if (_topology.bolts(i).comp().has_config()) {
+        const proto::api::Config& config = _topology.bolts(i).comp().config();
+        ConvertConfigToKVMap(config, retval);
+      }
+      return;
+    }
+  }
+}
+
+// Update component level config
+void TopologyConfigHelper::SetComponentConfig(proto::api::Topology* _topology,
+                                              const sp_string& _component_name,
+                                              const std::map<sp_string, sp_string>& _update) {
+  // We are assuming component names are unique and updating config for all instances
+  // with the specific component name.
+  for (sp_int32 i = 0; i < _topology->spouts_size(); ++i) {
+    proto::api::Component* comp = _topology->mutable_spouts(i)->mutable_comp();
+    if (comp->name() == _component_name && comp->has_config()) {
+      proto::api::Config* config = comp->mutable_config();
+      UpdateConfigFromKVMap(config, _update);
+    }
+  }
+
+  for (sp_int32 i = 0; i < _topology->bolts_size(); ++i) {
+    proto::api::Component* comp = _topology->mutable_bolts(i)->mutable_comp();
+    if (comp->name() == _component_name && comp->has_config()) {
+      proto::api::Config* config = comp->mutable_config();
+      UpdateConfigFromKVMap(config, _update);
+    }
+  }
+}
+
+// For every existing config, if it is in the new config, update the value
+void TopologyConfigHelper::UpdateConfigFromKVMap(proto::api::Config* _config,
+                                                 const std::map<sp_string, sp_string>& _update) {
+  for (sp_int32 i = 0; i < _config->kvs_size(); ++i) {
+    const std::string& key = _config->mutable_kvs(i)->key();
+    std::map<sp_string, sp_string>::const_iterator iter = _update.find(key);
+    if (iter != _update.end()) {
+      _config->mutable_kvs(i)->set_value(iter->second);
+    }
+  }
+}
+
 bool TopologyConfigHelper::GetBooleanConfigValue(const proto::api::Topology& _topology,
-                                                 const std::string& _config_name,
+                                                 const sp_string& _config_name,
```

Comment:
all new code should use std::string. sp_string is legacy

Reply:
got it. will do.